Detour/Source/DetourNavMeshQuery.cpp

```diff
@@ -998,106 +1106,207 @@ dtStatus dtNavMeshQuery::queryPolygons(const float* center, const float* halfExt
+		|| startPoses == 0
+		|| !endPos || !filter || maxPath <= 0 || !path || !pathCount)
 		return DT_FAILURE | DT_INVALID_PARAM;
 
-	if (startRef == endRef)
+	bool allStartRefsValid = true;
+	for (int i = 0; i < numStarts; ++i)
+		if (!m_nav->isValidPolyRef(startRefs[i]))
+			allStartRefsValid = false;
+
+	if (!allStartRefsValid)
+		return DT_FAILURE | DT_INVALID_PARAM;
+
+
 	{
-		path[0] = startRef;
-		*pathCount = 1;
-		return DT_SUCCESS;
+		bool startHasEnd = false;
+		for (int i = 0; i < numStarts; ++i)
+			if (startRefs[i] == endRef)
+				startHasEnd = true;
+
+		if (startHasEnd)
+		{
+			path[0] = endRef;
+			*pathCount = 1;
+			return DT_SUCCESS;
+		}
 	}
-	
-	m_nodePool->clear();
-	m_openList->clear();
-	
-	dtNode* startNode = m_nodePool->getNode(startRef);
-	dtVcopy(startNode->pos, startPos);
-	startNode->pidx = 0;
-	startNode->cost = 0;
-	startNode->total = dtVdist(startPos, endPos) * H_SCALE;
-	startNode->id = startRef;
-	startNode->flags = DT_NODE_OPEN;
-	m_openList->push(startNode);
-	
-	dtNode* lastBestNode = startNode;
-	float lastBestNodeCost = startNode->total;
-	
+
+	dtQueryPathState* ss;
+	if (!state)
+	{
+		dtAssert(m_nodePool);
+		dtAssert(m_openList);
+
+		ss = new (dtAlloc(sizeof(dtQueryPathState), DT_ALLOC_TEMP)) dtQueryPathState;
+		ss->m_nodePool = m_nodePool;
+		ss->m_openList = m_openList;
+	}
+	else
+	{
+		dtAssert(state->m_nodePool);
+		dtAssert(state->m_openList);
+		ss = state;
+	}
+
+
+	bool foundEarly = false;
+	dtNode* lastBestNode = 0;
+
+	if (!state || state->setStartRefs(startRefs, numStarts))
+	{
+		ss->m_nodePool->clear();
+		ss->m_openList->clear();
+		for (int i = 0; i < numStarts; ++i)
+		{
+			dtNode* startNode = ss->m_nodePool->getNode(startRefs[i]);
+			dtVcopy(startNode->pos, startPoses + (3 * i));
+			startNode->pidx = 0;
+			startNode->cost = 0;
+			startNode->total = dtVdist(startNode->pos, endPos) * H_SCALE;
+			startNode->id = startRefs[i];
+			startNode->flags = DT_NODE_OPEN;
+
+			ss->m_openList->push(startNode);
+		}
+
+		lastBestNode = ss->m_openList->top();
+	}
+	else
+	{
+		{
+			dtNode* endNode = state->m_nodePool->findNode(endRef, 0);
+			if (endNode && (endNode->flags & DT_NODE_CLOSED))
+			{
+				lastBestNode = endNode;
+				foundEarly = true;
+			}
+		}
+
+		// If we didn't find the end node, this means the end node has changed
+		// since the last call
+		// Thus we need to fixup the openset as the heuristic will have changed
+		if (!foundEarly)
+		{
+			if (state->m_openList->empty())
+			{
+				// Returning early here won't leak as we can't reach this location if
+				// we alloacted the temp search state
+				return DT_FAILURE;
+			}
+
+			dtNodeQueue* newOpenList = new (dtAlloc(sizeof(dtNodeQueue), DT_ALLOC_PERM)) dtNodeQueue(state->m_openList->getCapacity());
+			while(!state->m_openList->empty())
+			{
+				dtNode* node = state->m_openList->pop();
+				node->total = node->cost + dtVdist(node->pos, endPos) * H_SCALE;
+				newOpenList->push(node);
```

Comment:
Not sure if containers are used here. There are methods to push whole container.